src/server/editorServices.ts

```diff
@@ -855,65 +858,247 @@ namespace ts.server {
+                        WatchType.ConfigFileForInferredRoot, /*project*/ undefined, configFileName,
+                        (_filename, eventKind) => this.onConfigFileChangeForOpenScriptInfo(configFileName, eventKind)
+                    );
+                    this.logConfigFileWatchUpdate(configFileName, configFilePresenceInfo, ConfigFileWatcherStatus.UpdatedCallback);
+                }
+            }
+            else {
+                // There is no one tracking anymore. Remove the status
+                this.mapOfConfigFilePresence.delete(closedProject.canonicalConfigFilePath);
+            }
+        }
+
+        private logConfigFileWatchUpdate(configFileName: NormalizedPath, configFilePresenceInfo: ConfigFileExistence, status: ConfigFileWatcherStatus) {
+            if (this.logger.loggingEnabled()) {
+                const inferredRoots: string[] = [];
+                const otherFiles: string[] = [];
+                configFilePresenceInfo.trackingOpenFilesMap.forEach((value, key: Path) => {
+                    const info = this.getScriptInfoForPath(key);
+                    if (value) {
+                        inferredRoots.push(info.fileName);
+                    }
+                    else {
+                        otherFiles.push(info.fileName);
+                    }
+                });
+                const watchType = status === ConfigFileWatcherStatus.UpdatedCallback ||
+                    status === ConfigFileWatcherStatus.ReloadingFiles ||
+                    status === ConfigFileWatcherStatus.ReloadingInferredRootFiles ?
+                    (configFilePresenceInfo.configFileWatcher ? WatchType.ConfigFileForInferredRoot : WatchType.ConfigFilePath) :
+                    "";
+                this.logger.info(`ConfigFilePresence ${watchType}:: File: ${configFileName} Currently Tracking: InferredRootFiles: ${inferredRoots} OtherFiles: ${otherFiles} Status: ${status}`);
+            }
+        }
+
+        private closeConfigFileWatcherIfInferredRoot(configFileName: NormalizedPath, canonicalConfigFilePath: string,
+            configFilePresenceInfo: ConfigFileExistence, infoIsInferredRoot: boolean, reason: WatcherCloseReason) {
+            // Close the config file watcher if it was the last inferred root
+            if (infoIsInferredRoot &&
+                configFilePresenceInfo.configFileWatcher &&
+                !this.configFileExistenceTracksInferredRoot(configFilePresenceInfo)) {
+                this.closeFileWatcher(
+                    WatchType.ConfigFileForInferredRoot, /*project*/ undefined, configFileName,
+                    configFilePresenceInfo.configFileWatcher, reason
+                );
+                configFilePresenceInfo.configFileWatcher = undefined;
+            }
+
+            // If this was the last tracking file open for this config file, remove the cached value
+            if (!configFilePresenceInfo.trackingOpenFilesMap.size &&

```

Comment:
`configFilePresenceInfo.trackingOpenFilesMap.size` could be non-zero if there are `false` entries in there? But shouldn't a project that consists of only non-existent files be removed too?